pkg/storage/chunk/chunk_store_utils.go

```diff
@@ -105,26 +123,60 @@ type decodeResponse struct {
 // NewChunkFetcher makes a new ChunkFetcher.
 func NewChunkFetcher(cacher cache.Cache, cacheStubs bool, schema SchemaConfig, storage Client) (*Fetcher, error) {
 	c := &Fetcher{
-		schema:         schema,
-		storage:        storage,
-		cache:          cacher,
-		cacheStubs:     cacheStubs,
-		decodeRequests: make(chan decodeRequest),
+		schema:              schema,
+		storage:             storage,
+		cache:               cacher,
+		cacheStubs:          cacheStubs,
+		decodeRequests:      make(chan decodeRequest),
+		maxAsyncConcurrency: 16,
+		maxAsyncBufferSize:  1000,
+		stop:                make(chan struct{}, 1),
 	}
 
 	c.wait.Add(chunkDecodeParallelism)
 	for i := 0; i < chunkDecodeParallelism; i++ {
 		go c.worker()
 	}
 
+	// Start a number of goroutines - processing async operations - equal
+	// to the max concurrency we have.
+	c.asyncQueue = make(chan []Chunk, c.maxAsyncBufferSize)
+	for i := 0; i < c.maxAsyncConcurrency; i++ {
+		go c.asyncQueueProcessLoop()
+	}
+
 	return c, nil
 }
 
+func (c *Fetcher) writeBackCacheAsync(fromStorage []Chunk) error {
+	select {
+	case c.asyncQueue <- fromStorage:
+		return nil
+	default:
+		return errMemcachedAsyncBufferFull
+	}
+}
+
+func (c *Fetcher) asyncQueueProcessLoop() {
+	for {
+		select {
+		case fromStorage := <-c.asyncQueue:
+			cacheErr := c.writeBackCache(context.Background(), fromStorage)
+			if cacheErr != nil {
+				level.Warn(util_log.Logger).Log("msg", "could not store chunks in chunk cache", "err", cacheErr)
```

Comment:
I think this is a bit more clear

```suggestion
				level.Warn(util_log.Logger).Log("msg", "could not write fetched chunks from storage into chunk cache", "err", cacheErr)
```